src/app/clusters/network-commissioning-old/network-commissioning-old.cpp

```diff
@@ -0,0 +1,293 @@
+/*
+ *
+ *    Copyright (c) 2021 Project CHIP Authors
+ *    All rights reserved.
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+#include "network-commissioning.h"
+
+#include <cstring>
+#include <type_traits>
+
+#include <lib/support/CodeUtils.h>
+#include <lib/support/SafeInt.h>
+#include <lib/support/Span.h>
+#include <lib/support/ThreadOperationalDataset.h>
+#include <lib/support/logging/CHIPLogging.h>
+#include <platform/CHIPDeviceLayer.h>
+#include <platform/ConnectivityManager.h>
+#include <platform/internal/DeviceControlServer.h>
+
+#include <app-common/zap-generated/cluster-objects.h>
+
+#if CHIP_DEVICE_CONFIG_ENABLE_THREAD
+#include <platform/ThreadStackManager.h>
+#endif // CHIP_DEVICE_CONFIG_ENABLE_THREAD
+
+// Include DeviceNetworkProvisioningDelegateImpl for WiFi provisioning.
+// TODO: Enable wifi network should be done by ConnectivityManager. (Or other platform neutral interfaces)
+#if defined(CHIP_DEVICE_LAYER_TARGET)
+#define DEVICENETWORKPROVISIONING_HEADER <platform/CHIP_DEVICE_LAYER_TARGET/DeviceNetworkProvisioningDelegateImpl.h>
+#include DEVICENETWORKPROVISIONING_HEADER
+#endif
+
+// TODO: Configuration should move to build-time configuration
+#ifndef CHIP_CLUSTER_NETWORK_COMMISSIONING_MAX_NETWORKS
+#define CHIP_CLUSTER_NETWORK_COMMISSIONING_MAX_NETWORKS 4
+#endif // CHIP_CLUSTER_NETWORK_COMMISSIONING_MAX_NETWORKS
+
+using namespace chip;
+using namespace chip::app;
+
+namespace chip {
+namespace app {
+namespace Clusters {
+namespace NetworkCommissioning {
+
+constexpr uint8_t kMaxNetworkIDLen       = 32;
+constexpr uint8_t kMaxThreadDatasetLen   = 254; // As defined in Thread spec.
+constexpr uint8_t kMaxWiFiSSIDLen        = 32;
+constexpr uint8_t kMaxWiFiCredentialsLen = 64;
+constexpr uint8_t kMaxNetworks           = CHIP_CLUSTER_NETWORK_COMMISSIONING_MAX_NETWORKS;
+
+enum class NetworkType : uint8_t
+{
+    kUndefined = 0,
+    kWiFi      = 1,
+    kThread    = 2,
+    kEthernet  = 3,
+};
+
+struct ThreadNetworkInfo
+{
+    uint8_t mDataset[kMaxThreadDatasetLen];
+    uint8_t mDatasetLen;
+};
+
+struct WiFiNetworkInfo
+{
+    uint8_t mSSID[kMaxWiFiSSIDLen + 1];
+    uint8_t mSSIDLen;
+    uint8_t mCredentials[kMaxWiFiCredentialsLen];
+    uint8_t mCredentialsLen;
+};
+
+struct NetworkInfo
+{
+    uint8_t mNetworkID[kMaxNetworkIDLen];
+    uint8_t mNetworkIDLen;
+    uint8_t mEnabled;
+    NetworkType mNetworkType;
+    union NetworkData
```

Comment:
We have a type-safe variant in the sdk at src/lib/support/Variant.h.

Reply:
Oops.  Sorry.  I now realize that this is the old implementation.

Reply:
This is just a rename of the old network-commissioning, I don't want to touch it in this PR.